tests/test_state/test_storage_state.py

```diff
@@ -158,3 +301,294 @@ def test_clamp_energy_asserts_battery_traded_more_than_energy_per_slot(self, is_
+        storage_state = self._setup_storage_state_for_energy_origin_tracking()
+        available_energy_for_sell = 0
+        for energy in storage_state._used_storage_share:
+            available_energy_for_sell += energy.value
+        storage_state._track_energy_sell_type(energy=available_energy_for_sell)
+        if len(storage_state._used_storage_share) != 0:
+            assert isclose(
+                storage_state._used_storage_share[0].value, 0, abs_tol=FLOATING_POINT_TOLERANCE)
+        else:
+            assert len(storage_state._used_storage_share) == 0
+
+    def test_track_energy_sell_type_sell_more_energy_than_available(self):
+        storage_state = self._setup_storage_state_for_energy_origin_tracking()
+        available_energy_for_sell = 0
+        used_storage_share = storage_state._used_storage_share
+        for energy in used_storage_share:
+            available_energy_for_sell += energy.value
+        storage_state._track_energy_sell_type(energy=available_energy_for_sell + 0.1)
+        assert len(storage_state._used_storage_share) == 0
+
+    def test_track_energy_sell_type_sell_only_first_entry_completely(self):
+        storage_state = self._setup_storage_state_for_energy_origin_tracking()
+        energy_for_sale = 0.4
+        initial_registry_number = len(storage_state._used_storage_share)
+        first_origin = storage_state._used_storage_share[0].origin
+        storage_state._track_energy_sell_type(energy=energy_for_sale)
+        if len(storage_state._used_storage_share) != initial_registry_number - 1:
+            assert len(storage_state._used_storage_share) == initial_registry_number
+            assert storage_state._used_storage_share[0].origin == first_origin
+            assert isclose(
+                storage_state._used_storage_share[0].value, 0, abs_tol=FLOATING_POINT_TOLERANCE)
+        else:
+            assert len(storage_state._used_storage_share) == initial_registry_number - 1
+
+    def test_get_soc_level_default_values_and_custom_values(self):
+        storage_state = StorageState(initial_soc=100,
+                                     capacity=100)
+        current_time_slot, _, _ = self._initialize_time_slots()
+        storage_state.add_default_values_to_state_profiles([current_time_slot])
+        storage_state.activate(
+            slot_length=duration(minutes=15), current_time_slot=current_time_slot)
+        assert storage_state.get_soc_level(current_time_slot) == 1
+        storage_state.charge_history[current_time_slot] = 50
+        assert storage_state.get_soc_level(current_time_slot) == 0.5
+
+    def test_to_dict_keys_in_return_dict(self):
+        storage_state = StorageState()
+        current_time_slot, _, _ = self._initialize_time_slots()
+        storage_state.add_default_values_to_state_profiles([current_time_slot])
+        assert set(storage_state.to_dict(current_time_slot).keys()).issubset(SAMPLE_STATS.keys())
```

Comment:
The reason why I implemented this in such a way is because I imagined a scenario where an additional key is added to the stats dict on `to_dict()` method. 

I thought that if that happens, it should not break the test, because all of the previous keys (which might already have some use somewhere else) would be unaffected. If one of the currently existing keys is removed, then the test would indeed fail.

However, I did notice that if the test were to work as I intended it to, the subset condition should be the other way around.

```suggestion
        assert set(SAMPLE_STATS.keys()).issubset(storage_state.to_dict(current_time_slot).keys())
```

What do you think? Does this make sense to you? 
Or should we let the test break if any additional key is added to the stats dict, so that it becomes evident that the tests should be also updated/refactored?

Reply:
This way around it makes sense, indeed!